x/cash/dynamicfee.go

```diff
@@ -0,0 +1,235 @@
+package cash
+
+import (
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/gconf"
+	"github.com/iov-one/weave/x"
+)
+
+/*
+DynamicFeeDecorator is an enhanced version the basic FeeDecorator with
+better handling of tx errors, and ability to deduct/enforce
+app-specific fees.
+
+The business logic is:
+
+* If tx fee < min fee, or tx fee cannot be paid, reject with error.
+
+* Run the tx.
+
+* If tx errors, revert all tx changes and charge only the min fee
+
+TODO: * If tx succeeded, but requested a RequiredFee higher than paid fee,
+revert all tx changes and refund all but the min fee, returning an error.
+
+* If tx succeeded, and at least RequiredFee was paid, everything is committed
+and we return success
+
+It also embeds a checkpoint inside, so in the typical application stack:
+	cash.NewFeeDecorator(authFn, ctrl),
+	utils.NewSavepoint().OnDeliver(),
+can be replaced by
+	cash.NewDynamicFeeDecorator(authFn, ctrl),
+
+As with FeeDecorator, all deducted fees are send to the collector,
+whose address is configured via gconf package.
+*/
+type DynamicFeeDecorator struct {
+	auth    x.Authenticator
+	control FeeController
+	// these are cached values to not hit gconf on each read
+	collector weave.Address
+	minFee    *x.Coin
+}
+
+var _ weave.Decorator = DynamicFeeDecorator{}
+
+// NewDynamicFeeDecorator returns a DynamicFeeDecorator with the given
+// minimum fee, and all collected fees going to a default address.
+func NewDynamicFeeDecorator(auth x.Authenticator, control FeeController) DynamicFeeDecorator {
+	return DynamicFeeDecorator{
+		auth:    auth,
+		control: control,
+	}
+}
+
+// Check verifies and deducts fees before calling down the stack
+func (d DynamicFeeDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx,
+	next weave.Checker) (weave.CheckResult, error) {
+
+	// prepare fees for subtransaction -- shared in Check and Deliver
+	var res weave.CheckResult
+	fee, payer, cache, err := d.prepare(ctx, store, tx)
+	if err != nil {
+		return res, err
+	}
+
+	// read config
+	minFee := d.getMinFee(store)
+
+	// do subtransaction in a function for easier error handling
+	res, err = func() (weave.CheckResult, error) {
+		// shadow with local variables...
+		err := d.maybeTakeFee(cache, payer, fee)
+		if err != nil {
+			return weave.CheckResult{}, err
+		}
+		res, err := next.Check(ctx, cache, tx)
+		// TODO: check RequiredFee here and return an error if insufficient
+		return res, err
+	}()
+
+	// on error, rollback, then take the minfee from the store
+	if err != nil {
+		cache.Discard()
+		// if this fails, we aborted early above, we can just ignore return value
+		// this is 2 ops, not 1, for errors, but done to optimize the success case to use 1 not 2
+		d.maybeTakeFee(store, payer, minFee)
```

Comment:
I mean, if I send with 0 fee, it will be rejected, I could do two checks in the beginning, but to use only one send in success case, we need the second send in error case. It is not a problem if this errors, and it will be rejected in Check and should never make it to a block anyway